optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,287 @@
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, _ = self._hot_cache[trial_id]
+                return cached_trial.number
+
+        return self._storage.get_trial_number_from_id(trial_id)
+
+    def get_best_trial(self, study_id: int) -> FrozenTrial:
+
+        return self._storage.get_best_trial(study_id)
+
+    def get_trial_param(self, trial_id: int, param_name: str) -> float:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, _ = self._hot_cache[trial_id]
+                return cached_trial.distributions[param_name].to_internal_repr(
+                    cached_trial.params[param_name]
+                )
+
+        return self._storage.get_trial_param(trial_id, param_name)
+
+    def set_trial_value(self, trial_id: int, value: float) -> None:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                cached_trial.value = value
+                updates.value = value
+                if trial_id in self._dirty_trials:
+                    self._dirty_trials.remove(trial_id)
+                self._update_storage(trial_id)
+                return
+
+        self._storage.set_trial_value(trial_id, value)
+
+    def set_trial_intermediate_value(
+        self, trial_id: int, step: int, intermediate_value: float
+    ) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                if step in cached_trial.intermediate_values:
+                    return False
+                cached_trial.intermediate_values[step] = intermediate_value
+                updates.values[step] = intermediate_value
+                if trial_id in self._dirty_trials:
+                    self._dirty_trials.remove(trial_id)
+                self._update_storage(trial_id)
+                return True
+
+        return self._storage.set_trial_intermediate_value(trial_id, step, intermediate_value)
+
+    def set_trial_user_attr(self, trial_id: int, key: str, value: Any) -> None:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                self._dirty_trials.add(trial_id)
+                cached_trial.user_attrs[key] = value
+                updates.user_attrs[key] = value
+                return
+
+        self._storage.set_trial_user_attr(trial_id, key, value)
+
+    def set_trial_system_attr(self, trial_id: int, key: str, value: Any) -> None:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                self._dirty_trials.add(trial_id)
+                cached_trial.system_attrs[key] = value
+                updates.system_attrs[key] = value
+                return
```

Comment:
Since `SuccessiveHalvingPruner` accesses to the system attributes of other trials (including RUNNING ones), it seems needed to delete this block to keep compatibility with the existing behavior.
https://github.com/optuna/optuna/blob/5c07d0afffeb506866860fa2ca21a71a4199c075/optuna/pruners/successive_halving.py#L221

Reply:
I fixed this issue.